### PR TITLE
[pruner] state kv pruner

### PR DESCRIPTION
**Pull request by lightmark**
### Description
prune state value as part of ledger pruner.

also fix the restore of empty tree.

### Test Plan
ut

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3106)

### CODE REVIEW

File: storage/aptosdb/src/test_helper.rs [hunk 1]
```diff
@@ -18,8 +21,51 @@ use aptos_types::{
...
+    input: impl Iterator<Item = (StateKey, Option<StateValue>)>,
+    first_version: Version,
+) -> HashValue {
+    let mut root_hash = *SPARSE_MERKLE_PLACEHOLDER_HASH;
+    for (i, (key, value)) in input.enumerate() {
+        let value_state_set = vec![(key, value)].into_iter().collect();
+        let jmt_updates = jmt_updates(&value_state_set);
+        let version = first_version + i as Version;
+        root_hash = store
+            .merklize_value_set(
```
Comment (msmouse) [inline storage/aptosdb/src/test_helper.rs:54]: need to make it on for feature=fuzzing as well.

File: storage/aptosdb/src/test_helper.rs [hunk 2]
```diff
@@ -18,8 +21,51 @@ use aptos_types::{
     proptest_types::{AccountInfoUniverse, BlockGen},
 };
 use executor_types::ProofReader;
+use proptest::sample::Index;
 use proptest::{collection::vec, prelude::*};
 use scratchpad::SparseMerkleTree;
+use storage_interface::{jmt_update_refs, jmt_updates};
+
+prop_compose! {
+    pub fn arb_state_kv_sets(key_universe_size: usize, update_set_size_per_version: usize, num_versions: usize)(keys in vec(any::<StateKey>(), key_universe_size), input in vec(vec((any::<Index>(), any::<Option<StateValue>>()), 1..update_set_size_per_version), 1..num_versions)) -> Vec<Vec<(StateKey, Option<StateValue>)>> {
```
Comment (msmouse) [inline storage/aptosdb/src/test_helper.rs:30]: format a bit?

Comment (msmouse) [reply storage/aptosdb/src/test_helper.rs:30]: "update_set_size_per_version" -> "max_update_set_size"
"num_versions" -> "max_versions"

File: storage/aptosdb/src/state_store/mod.rs [hunk 3]
```diff
@@ -649,6 +649,28 @@ impl StateStore {
             expected_root_hash,
         )?))
     }
+
+    /// Prune the stale state value schema generated between a range of version in (begin, end]
```
Comment (msmouse) [inline storage/aptosdb/src/state_store/mod.rs:653]: "[begin, end)"

File: storage/aptosdb/src/pruner/state_store/test.rs [hunk 4]
```diff
@@ -359,3 +361,65 @@ fn test_worker_quit_eagerly() {
...
+
+fn verify_state_value<'a, I: Iterator<Item = (&'a StateKey, &'a Option<StateValue>)>>(
+    kvs: I,
+    version: Version,
+    state_store: &Arc<StateStore>,
+    pruned: bool,
+) {
+    for (k, v) in kvs {
+        let v_from_db = state_store.get_state_value_by_version(k, version).unwrap();
+        assert_eq!(&v_from_db, if pruned { &None } else { v });
```
Comment (msmouse) [inline storage/aptosdb/src/pruner/state_store/test.rs:423]: 👍

**Review by msmouse — APPROVED** [review]

File: storage/aptosdb/src/state_merkle_db.rs [hunk 5]
```diff
@@ -182,7 +183,12 @@ impl TreeReader<StateKey> for StateMerkleDb {
         let mut iter = self.iter::<JellyfishMerkleNodeSchema>(Default::default())?;
         iter.seek_to_first();
         let version = match iter.next().transpose()? {
-            Some((node_key, _node)) => node_key.version(),
+            Some((node_key, node)) => {
+                if node.node_type() == NodeType::Null {
+                    return Ok(None);
+                }
+                node_key.version()
+            }
```
Comment (zcchahaha) [inline storage/aptosdb/src/state_merkle_db.rs:191]: For my own understanding, why do we need to change this? Is this `fix the restore of empty tree` you mentioned?

Comment (lightmark) [reply storage/aptosdb/src/state_merkle_db.rs:191]: yes

Comment (lightmark) [reply storage/aptosdb/src/state_merkle_db.rs:191]: Otherwise another test will fail.

File: storage/aptosdb/src/state_store/mod.rs (hunk 3 above)
Comment (lightmark) [reply storage/aptosdb/src/state_store/mod.rs:653]: It is confusing because stale_since_version will be `end` inclusive.

**Review by zcchahaha — APPROVED** [review]

**Pull request merged**